Allow users to set a duration time in minutes for a poll due date

masch opened the pull request:
Adds a menu settings option to set a specific due time duration in minutes from current datetime. 

For example, if a user set 25 minutes of due time duration and the current time is 15/10/2022 10:10 HS it will set the due datetime on 15/10/2022 10:35 HS. 
The duration time is added on the due datetime data, so any edge case of a date is cover by the addition.

masch committed:
Allow users to set a duration time in minutes for a poll due date

Adds a menu settings option to set a specific due time duration in
minutes from current datetime. For example, if a user set 25 minutes
of due time duration and the current time is 15/10/2022 10:10 HS
it will set the due datetime on 15/10/2022 10:35 HS.
The duration time is added on the due datetime data, so any
edge case of a date is cover by the addition.

masch commented:
Hi @Nukesor, I changed the PR's status without any change in order to ask your opinion about the changes.
Thanks!

masch commented on pollbot/models/poll.py:
```diff
@@ -152,9 +152,9 @@ def get_date_option(self, check_date: date) -> Optional[Option]:
         """Get the formatted date."""
         if self.european_date_format:
-            return self.due_date.strftime("%d.%m.%Y %H:%M UTC")
+            return self.due_date.strftime("%d.%m.%Y %H:%M")
```
I think is this case, the time displayed is in the corresponding unit time of the poll, it's not UTC, is it?

Nukesor replied:
All datetimes on this codebase are actually `timezone without time zone`.
Since the server has a UTC time, this is UTC.

The thing is, that telegram usually sends UTC timestamps, and it's extremely hard to reliably find out where a user is from.

Due to this, I went for the easiest approach, i.e. working with UTC.

masch replied:
Sorry, I didn't express it correct. Let me give you more information.
You are right, all date times on the codebase are in UTC format, but the date time displayed on the due date is not in UTC format, it displays the date time with the location of the telegram's user. Let me give you an example:
In my country, we use `UTC -3` format, so now in my country it's 19:15 HS `UTC-3` so in `UTC` time it's ` 22:15 hs`.
If I create a poll with 30 minutes on due time duration, the due date time will be: `19:45 hs on UTC-3` and `22:45 hs on UTC`
The message displayed on my telegram is:
![image](https://user-images.githubusercontent.com/53308354/199846489-3e36ed14-66dc-43ad-8f43-da28af68749d.png)
If you see the picture, the date time displayed is `UTC-3` (19:45) and not `UTC` (22:45)
So that why I think that we use remove the text `UTC`.
Another thing to mention, I don't if it's right, but the due date time in the database is saved on `UTC -3` and created at is saved in `UTC`. Here is the output:

```
 id | locale  |         created_at         |         updated_at         |          due_date          |     next_notification      
----+---------+----------------------------+----------------------------+----------------------------+----------------------------
  2 | English | 2022-11-03 22:15:19.133202 | 2022-11-03 22:15:12.590988 | 2022-11-03 19:45:04.991982 | 2022-11-03 19:45:04.991982
````
The poll works fines, it's finished on my local time, but I don't if we should save the date time on UTC format like the other fields.
Let me know what do you think about the example.

Nukesor replied:
This is because you're running your database on a server/pc that has the same local time as you.
Once you run this on a server in a different timezone, you'll see the expected time shift.

The official Ultimate Pollbot server runs on a server that has a UTC timezone -> All times are UTC.

If we're going to deploy this code on the real instance, users will see a local time i.e. 19:45, even though the time will actually be 19:45 UTC.

Nukesor replied:
I admit that this is quite hacky, but back when I started this project, I didn't spend a single thought on running this for a global userbase :D  

In general, it would be best to refactor this whole system.

But for this to properly work, we would still have to keep track of user's locales, which isn't easy to automate for bots (at least it wasn't last time I checked). We would have to add a new field and a new form that allows users to pick their timezone

Nukesor commented:
@masch Could you rebase onto main?

I introduced a proper CI pipeline, code style checks and a streamlined setup.

masch commented:
> @masch Could you rebase onto main?
> 
> I introduced a proper CI pipeline, code style checks and a streamlined setup.

I rebased it. In order to fix code style, should I run `just format` and `just lint` and push the changes or do i need to do something else?

Nukesor commented:
`just format` should do the job, `just lint` to see if there's anything left that hasn't been corrected by the format command :)

masch commented:
> `just format` should do the job, `just lint` to see if there's anything left that hasn't been corrected by the format command :)

Ok, I run it and fixed all the suggestion. The job is running now, Do I need to run something else?

Nukesor commented:
Nope, Now it's time for me to check out your code and do a proper review :D

Sorry for the whole CI stuff. I haven't worked on this for such a long time, that I had to work out how to setup everything.
Took the chance and cleaned up the project.

Nukesor commented on pollbot/telegram/message_handler.py:
```diff
@@ -249,3 +256,38 @@ def handle_user_option_addition(
+    if not text.isnumeric():
+        return i18n.t("creation.error.limit_bigger_zero", locale=user.locale)
+
+    "Set 1 week in minutes as arbitrary maximum relative value"
+    maximum_relative_minutes = 60 * 24 * 7
+    relative_minutes = int(text)
+    if relative_minutes < 1 or relative_minutes > maximum_relative_minutes:
+        return i18n.t(
+            "creation.error.limit_between",
+            locale=user.locale,
+            limit=maximum_relative_minutes,
+        )
+
+    poll_due_date = datetime.now()
+    next_poll_due_date = poll_due_date + relativedelta(minutes=relative_minutes)
+    poll.set_due_date(next_poll_due_date)
+
+    chat.send_message(
+        text=get_due_time_duration_text(poll),
+        reply_markup=get_add_due_time_keyboard(poll),
+        parse_mode="markdown",
+    )
```
To be honest, I didn't expect this to be a relative number input in minutes.
This is super hard to properly convey to users, especially to users that aren't used to work with time that way.

I imagined this to be built into the already existing keyboard.
Something like

```
DATEPICKER HERE
[       Hours         ]
[+4][+1][-1][-4]
[       Minutes      ]
[+15][+5][+1][-1][-5][-15]
```

Nukesor reviewed (changes requested):
I think the current solution on how to solve this isn't user-friendly enough yet.

We already have the due time datepicker and should definitely build on top of this!


Furthermore, I decided to lock down this repository for maintenance only.
I can no longer find any motivation to actively work on this huge piece of Python :D.
This PR showed me, that working on old projects takes a lot of time, especially if one no longer works with a specific programming language.

So this will be the very last PR I'll accept on this project :laughing:

semaf commented:
Sorry that I hook up here but didnt find any other contact option to you. @Nukesor 

The bots reply:

I would like to see a specific feature being added to the bot
Check out the official feature tracking board: https://trello.com/b/L3XfSRkN/ultimate-poll-bot
Some features won't be implemented, as they don't match the guidelines of the project.
Some frequently requested features are already planned and are waiting for somebody to code them.
If your feature isn't listed on the trello board, feel free create an Issue on GitHub (https://github.com/Nukesor/ultimate-poll-bot).

Issue is disabled.

Nukesor commented:
Thanks for pointing that out, I'll remove this notice from the bot, as it's currently locked down and is no longer actively developed or maintained

semaf commented:
@Nukesor Its out of topic but the bot is down since yesterday. Maybe due your release?

Nukesor commented:
Ping @masch Are you planning to continue working on this :)?

Nukesor commented:
Closed due to inactivity
